protobuf/src/main/java/io/grpc/protobuf/reflection/ProtoServiceDescriptor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.protobuf.reflection;
+
+import com.google.protobuf.Descriptors.FileDescriptor;
+
+import io.grpc.AbstractServiceDescriptor;
+import io.grpc.MethodDescriptor;
+import io.grpc.ServiceDescriptor;
+
+import java.util.Collection;
+
+/**
+ * Descriptor for a protobuf-backed service. This adds a method to retrieve the protobuf file
+ * descriptor, but otherwise only wraps a {@link ServiceDescriptor} instance.
+ */
+public abstract class ProtoServiceDescriptor extends AbstractServiceDescriptor {
```

Comment:
I found this in one of the collapsed comments:
> Alternatively, instead of pulling an abstract class out of ServiceDescriptor, I could add a @Nullable SomeInterface getAttachedObject() method to the existing ServiceDescriptor class. Non-protobuf code generation would ignore this optional field on the ServiceDescriptor, but the protobuf codegen would attach a new ProtoFileDescriptorWrapper extends SomeInterface instance that has an accessor method to the underlying proto file descriptor for reflection. This would avoid the need for AbstractServiceDescriptor altogether.

Yeah, that is what I was expecting. I'm not sure if we gain much by SomeInterface; it may be good from a documentation standpoint, but I can't think of any methods that should be on it.